[xqtann] iP

xqtann opened the pull request:
### Pikappi
Pikappi is a **text-based** chatbot that is *easy to use* and learn. 

It consists of the following commands:
- todo <task>
- deadline <task> /by <deadline>
- event <task> /from <time> /to <time>
- mark <taskNumber>
- unmark <taskNumber>
- delete <taskNumber>
- list
- find <keyword>

You can try it out by:
1. download it from [here](https://github.com/xqtann/ip/releases/tag/A-Jar).
2. add your tasks to Pikappi.
3. organise and manage your tasks **efficiently**! 👍 

This is a sample `main` method in Java:
``` java
 public static void main(String[] args) throws PikappiException {
        ui.greet();
        boolean isExit = false;


        while (!isExit) {
            String command = ui.readCommand();
            ui.showLine();
            try {
                Command c = parser.parse(command);
                c.execute(tasks, ui, storage);
            } catch (PikappiException e) {
                ui.showErrorMessage(e.getMessage());
            }
            isExit = parser.isExit();
            ui.showLine();
       }
    }
```

Have you done the following?
- [ ] read through my PR description
- [x] tried out **Pikappi**

Lastly, heres a quote of the day.
> _"Simplicity is the soul of efficiency."_

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

xqtann committed:
add Pikappi.java

xqtann committed:
added echo commands

xqtann committed:
added Add, List commands

xqtann committed:
change to oop

xqtann committed:
add Task.java

xqtann committed:
add mark and unmark commands

xqtann committed:
add DeadlineTask, EventTask, TodoTask

xqtann committed:
add functions of each task type

xqtann committed:
add text ui testing

xqtann committed:
add pikappiException

xqtann committed:
change array to arraylist

xqtann committed:
add delete command

xqtann committed:
update text ui test

xqtann committed:
use enums for tasktype

xqtann committed:
Add loadTasks method to read saved tasks from data/pikappi.txt

xqtann committed:
Add saveTasks method to save tasks to data/pikappi.txt

xqtann committed:
Merge branch 'branch-Level-7'

xqtann committed:
Add exception for corrupted file

xqtann committed:
Merge branch 'branch-Level-7'

xqtann committed:
Add formatted date for tasks

xqtann committed:
Add time format for tasks

xqtann committed:
Merge branch 'branch-Level-8'

xqtann committed:
Add Storage class for more OOP

xqtann committed:
Add TaskList for OOP

xqtann committed:
Add Ui class for more OOP

xqtann committed:
Add Parser for more OOP

xqtann committed:
Add Command classes for OOP

xqtann committed:
Add switch case statements for commands

xqtann committed:
Add and classify packages

xqtann committed:
Merge branch 'master' into add-gradle-support

xqtann committed:
Add gradle

xqtann committed:
Add test cases for JUnit testing

xqtann committed:
Add Javadoc to non-private classes and methods

xqtann committed:
Change code to follow coding standards

xqtann committed:
Add find command to find tasks

xqtann committed:
Change and organise code for more OOP

xqtann committed:
Merge branch 'branch-A-CodingStandard'

xqtann committed:
Fix merge conflict

xqtann committed:
Fix minor bugs from merge conflict

xqtann committed:
Add JavaFX initial files

xqtann committed:
Add checkstyle checks

xqtann committed:
Merge branch 'branch-A-CheckStyle'

xqtann committed:
Merge branch 'master' into branch-Level-10

yixianggg commented on src/main/java/pikappi/Ui.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Bids farewell to the user when the program ends.
+     */
+    public void goodbye() {
```
Perhaps use a better method name (verb) like sayGoodbye?

yixianggg commented on src/main/java/pikappi/command/Command.java:
```diff
@@ -0,0 +1,19 @@
+package pikappi.command;
```
I like how there are separate packages within the pikappi package to distinguish command classes, task classes and the exception class!

yixianggg commented on src/main/java/pikappi/Parser.java:
```diff
@@ -0,0 +1,90 @@
+            }
+        default:
+            throw new PikappiException("Pika..?? I don't understand..");
+        }
```
Good job for following the coding standards regarding the switch case! 👍

yixianggg commented on src/main/java/pikappi/Pikappi.java:
```diff
@@ -0,0 +1,34 @@
+                ui.showErrorMessage(e.getMessage());
+            }
+            isExit = parser.isExit();
+            ui.showLine();
```
Should this be extracted out and let there be a Main class that calls upon the Pikappi class?

yixianggg reviewed (commented):
LGTM. Just some small nits to fix. Well done!

xqtann committed:
Add GUI to Pikappi and fix minor DateTime bug

xqtann committed:
Merge branch 'branch-Level-10'

xqtann committed:
Add Varargs in FindCommand

xqtann committed:
Merge branch 'branch-A-Varargs'

xqtann committed:
Change minor styling of GUI

xqtann committed:
Add help command

xqtann committed:
Add assertions

Assertions are not included in any of the files.

This can cause possible bugs in code where assumptions have been made.

Adding assertions can allow runtime to verify the defined assumptions by
us.

Let's add assertions statements at places where assumptions are made.

Using assertions this way can reduce unexpected bugs.

xqtann committed:
Add enableAssertions in build.gradle

xqtann committed:
Change code to adhere to code quality rules.

Code is difficult to read and follow, with minimal adherance to code
quality rules.

Improvement in code quality can enhance code readability, and understand
more easily.

Let's refactor the code to make it more explicit and understandable,
while keeping it simple.

Code refactoring can improve code readability and make code more neat,
simple and easy to follow and read.

xqtann committed:
TaskList.java: add streams in findTask method

findTask method uses nested for loop to filter out tasks with specified keywords.

It can cause confusion and harder to read code with the nested loop.

Let's use Java Streams to simplify the nested loop.

Using streams can improve performance and code readability.

yu-sutong commented on src/main/java/pikappi/exception/PikappiException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represents an exception that occurs in Pikappi.
+ */
+public class PikappiException extends Exception {
```
Should there be a more specific Exception to catch to avoid Pokemon Exception Handling?

yu-sutong commented on src/main/java/pikappi/task/Task.java:
```diff
@@ -0,0 +1,80 @@
+package pikappi.task;
```
I like how there are separate packages within the pikappi package to distinguish between the different class functionalities.

yu-sutong commented on src/main/java/pikappi/command/Command.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Represents a command that can be executed by the user.
+ */
+public abstract class Command {
```
I like how you have a Command class and subclasses that extend from it. You went the extra mile to make your code clearer.

yu-sutong reviewed (commented):
LGTM.

xqtann committed:
Merge pull request #1 from xqtann/branch-A-Assertions

Add assertions

xqtann committed:
Resolve merge conflicts

xqtann committed:
Merge pull request #2 from xqtann/branch-A-CodeQuality

Change code to adhere to code quality rules.

xqtann committed:
Resolve merge conflicts

xqtann committed:
Merge branch 'master' into branch-A-Streams

xqtann committed:
Merge pull request #3 from xqtann/branch-A-Streams

TaskList.java: add streams in findTask method

xqtann committed:
Change stored tasks data

xqtann committed:
Merge branch 'branch-A-Assertions'

xqtann committed:
Add gradle.yml file for CI

xqtann committed:
Update gradle.yml

xqtann committed:
Update gradle.yml

xqtann committed:
Update gradle.yml

xqtann committed:
Add Javadoc for MainWindow.java

xqtann committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

xqtann committed:
Add sort command

User cannot sort tasks in tasklist.

Being able to sort tasks in the list can help maintain a neater
tasklist, especially when list is long.

Let's add a sort command where users can sort by task description, task
type, status, deadline, and event start time.

Sorting tasks allow users to maintain a neater tasklist and find tasks
more easily.

xqtann committed:
Merge branch 'branch-C-Sort'

xqtann committed:
Add Javadocs in SortCommand.java

xqtann committed:
Change GUI

xqtann committed:
Add error handling for multiple parameters in command

xqtann committed:
EventTask: Add error handling for start date later than end date

xqtann committed:
Add more JUnit test cases

xqtann committed:
TaskList addTask: add error handling for duplicate tasks

xqtann committed:
Add JUnit test case for duplicate tasks

xqtann committed:
Add Pikappi README User Guide and Ui.png

xqtann committed:
Merge branch 'branch-A-UserGuide'

xqtann committed:
Add javadocs and minor change to code style

xqtann committed:
Update README

xqtann committed:
Update README
